src/index.js

```diff
@@ -1,130 +1,242 @@
+	integrationSpecPaths: [ "./spec/integration" ],
+	specs: [ "**/*.spec.js" ],
+	watchPaths: [ "./src/**/*", "./spec/**/*", "./resource/**/*" ],
+	sourcePaths: [ "./src", "./resource" ],
+	sources: [ "**/*.js" ],
+	coverageHtml: "./coverage/lcov-report/index.html",
+	exclude: [ "!node_modules*" ],
+	jscsCfgPath: ".jscsrc"
+};
+var jshint = require( "gulp-jshint" );
+var jscs = require( "gulp-jscs" );
+var gutil = require( "gulp-util" );
+var gulpChanged = require( "gulp-changed" );
+var stylish = require( "jshint-stylish" );
 
-function groklate( ngrok, subdomain, port, token ) {
-	return when.promise( function( resolve, reject ) {
-		ngrok.connect( {
-			port: port,
-			subdomain: subdomain,
-			authtoken: token
-		}, function( err, url ) {
-				if ( err ) {
-					reject( err );
-				} else {
-					resolve( url );
-				}
-			} );
-	} );
+function permuPath( dirs, globs ) {
+	return _.reduce( globs, function( accum, glb ) {
+		Array.prototype.push.apply(
+			accum,
+			_.map( dirs, function( dir ) {
+				return path.join( dir, glb );
+			} )
+		);
+		return accum;
+	}, [] );
 }
 
-function init() {
-	gulp.task( 'restartProcesses', function( cb ) {
-		processhost.restart()
-			.then( function() {
-				cb();
-			}, function( err ) {
-					cb( err );
-				} );
-	} );
-}
+module.exports = function( gulpRef, cfg ) {
+	var gulp = gulpRef;
+	var options = _.defaults( cfg || {}, defaults );
+	var processesDefined = false;
 
-function runSpecs( specPath, coverage ) {
-	var args = Array.prototype.slice.call( arguments, 2 );
-	if ( args.length === 0 ) {
-		args = [ defaultSpecs ];
+	function runSpecs( testCfg ) {
+		return gulp.src( testCfg.specs, { read: false } )
+			.pipe( mocha( {
+				R: "spec",
+				istanbul: testCfg.coverage ? { x: testCfg.coverageExclude } : false
+			} ) );
 	}
-	var joinArgs = _.flatten( args );
-	var specs = _.reduce( joinArgs, function( acc, spec ) {
-		_.each( specPath, function( p ) {
-			acc.push( path.join( p, spec ) );
+
+	function test( specs ) {
+		var s = _.isArray(specs) ? specs: [ specs ];
+		return runSpecs( {
+			specs: s
 		} );
-		return acc;
-	}, [] );
-	var specExclude = _.map( specPath, function ( p ) {
-		return path.join( p, '**/*' );
-	} );
-	return gulp
-		.src( specs, { read: false } )
-		.pipe( mocha( { R: 'spec', istanbul: coverage ? {
-			x: specExclude
-		} : false } ) );
-}
+	}
 
-function setupProcess( processName, opts ) {
-	processesDefined = true;
-	return processhost.startProcess( processName, {
-		command: opts.command || opts.cmd || 'node' + cmdPostfix,
-		args: opts.arguments || opts.args || [],
-		stdio: opts.stdio || 'inherits',
-		restart: opts.restart
-	} );
-}
+	function testOnce( specs ) {
+		return test( specs )
+			.on( "end", process.exit.bind( process, 0 ) )
+			.on( "error", process.exit.bind( process, 100 ) );
+	}
 
-function setupProcesses( opts ) {
-	processesDefined = true;
-	return processhost.setup( opts );
-}
+	function testAll( opt ) {
+		var _opt = opt || options;
```

Comment:
Should this be merging options (e.g. `_.defaults(opt || {}, options)`) rather than just selecting one over the other? Applies to several other functions as well.